### PR TITLE
Option to select agents whose status will be updated in the next inventory according to their current status.

### PR DESCRIPTION
**Pull request by Lainow**
Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | !30987

Added a multiple-choice drop-down list to select the agents whose state will be modified the next time the Cleanoldagents cron job is run.
image

### CODE REVIEW

File: templates/pages/login.html.twig [hunk 1]
```diff
@@ -34,97 +34,97 @@
...
 
 {% block javascript_block %}
-<script type="text/javascript">
-   $(function () {
+	<script type="text/javascript">
+		$(function () {
 $('#login_name').focus();
 });
-</script>
+	</script>
```
Comment (stonebuzz) [inline templates/pages/login.html.twig:129]: wrong indent ?

File: tests/functional/Agent.php [hunk 2]
```diff
@@ -500,5 +500,41 @@ public function testStaleActions()
...
+        );
+
+        //run crontask
+        $task = new \CronTask();
+        $this->integer(\Agent::cronCleanoldagents($task))->isIdenticalTo(1);
+
+        //check item has been updated
+        $this->boolean($item->getFromDB($item->fields['id']))->isTrue();
+        $this->integer($item->fields['is_deleted'])->isIdenticalTo(1);
+        $this->integer($item->fields['states_id'])->isIdenticalTo($another_states_id);
```
Comment (stonebuzz) [inline tests/functional/Agent.php:538]: another check to test the ```all if empty``` option should be added

File: src/Agent.php [hunk 3]
```diff
@@ -881,17 +881,30 @@ public static function cronCleanoldagents($task = null)
...
-                                'id'        => $item->fields['id'],
-                                'states_id' => $config['stale_agents_status'],
-                                'is_dynamic' => 1
-                            ];
-                            if ($item->update($input)) {
-                                $task->addVolume(1);
-                                $total++;
+                            $shouldUpdate = false;
+                            if (
+                                (isset($config['old_stale_agents_status']) && $config['old_stale_agents_status'] != json_encode([0]))
```
Comment (stonebuzz) [inline src/Agent.php:886]:
```suggestion
                                (isset($config['previous_stale_agents_status']) && $config['previous_stale_agents_status'] != json_encode([0]))
```

Same for all occurence of ```old_stale_agents_status```

**Review by stonebuzz — CHANGES_REQUESTED** [review]

File: src/Inventory/Conf.php [hunk 4]
```diff
@@ -939,9 +939,13 @@ function changestatus() {
                 if ($('#dropdown_stale_agents_action$rand').val() != 0) {
                 $('#blocaction1').show();
                 $('#blocaction2').show();
+                $('#blocaction3').show();
+                $('#blocaction4').show();
                 } else {
                 $('#blocaction1').hide();
                 $('#blocaction2').hide();
+                $('#blocaction3').hide();
+                $('#blocaction4').hide();
```
[the remark this answers is not in the document]
Comment (Lainow) [reply src/Inventory/Conf.php:948]: I didn't succeed in putting it in one but in 2 it works.

File: templates/pages/login.html.twig [hunk 5]
```diff
@@ -127,4 +127,4 @@
 $('#login_name').focus();
 });
 </script>
-{% endblock %}
+{% endblock %}
```
Comment (trasher) [inline templates/pages/login.html.twig:130]: Please avoid changes on unrelated files

File: tests/functional/Agent.php [hunk 6]
```diff
@@ -500,5 +500,75 @@ public function testStaleActions()
         $this->boolean($item->getFromDB($item->fields['id']))->isTrue();
         $this->integer($item->fields['is_deleted'])->isIdenticalTo(1);
         $this->integer($item->fields['states_id'])->isIdenticalTo($states_id);
+
+        //create another new status
```
Comment (trasher) [inline tests/functional/Agent.php:504]: As far as I understand the original need; we should ensure state is not changed on agent that have a state not configured to be changed

File: templates/pages/login.html.twig (hunk 5 above)
Comment (Lainow) [reply templates/pages/login.html.twig:130]: This change doesn't make sense, it tells me I've added something, but it's already there.

Comment (trasher) [reply templates/pages/login.html.twig]: Fixed in 1190d286e236a83d46c8caa1cf00216794980e20

File: src/State.php [hunk 7]
```diff
@@ -655,6 +655,36 @@ public static function getDisplayConditionForAssistance(): array
         ];
     }
 
+    public static function getStateNotForInventory()
```
Comment (trasher) [inline src/State.php:658]: Need to be tested

File: src/State.php [hunk 8]
```diff
@@ -655,6 +655,36 @@ public static function getDisplayConditionForAssistance(): array
...
+                if ($state['is_visible_' . strtolower($inv_type)] === 0) {
+                    $output[$state['id']] = $state['id'];
+                    break;
+                }
+            }
+        }
+        return $output;
+    }
+
+    public static function getStateList()
```
Comment (trasher) [inline src/State.php:677]: Need to be tested

File: src/Agent.php [hunk 9]
```diff
@@ -881,17 +881,30 @@ public static function cronCleanoldagents($task = null)
...
-                                'id'        => $item->fields['id'],
-                                'states_id' => $config['stale_agents_status'],
-                                'is_dynamic' => 1
-                            ];
-                            if ($item->update($input)) {
-                                $task->addVolume(1);
-                                $total++;
+                            $shouldUpdate = false;
+                            if (
+                                (isset($config['previous_stale_agents_status']) && $config['previous_stale_agents_status'] != json_encode([0]))
```
Comment (trasher) [inline src/Agent.php:886]: More easy to read, drop extra useless parenthesis:
```suggestion
                                isset($config['previous_stale_agents_status'])
                                && $config['previous_stale_agents_status'] != json_encode([0])
```

File: src/Inventory/Conf.php [hunk 10]
```diff
@@ -935,38 +935,51 @@ public function showConfigForm()
...
+                    'multiple'  => true,
+                    'used'      => State::getStateNotForInventory(),
+                ]
+            );
+            echo "</td>";
+            echo "</tr>";
+
+            echo "<tr class='tab_bg_1' style='display:none' id='bloc_status_action2'><td colspan=2></td>";
+            echo "<td>";
+            echo \State::createTabEntry(__('Status changed'), 0, \State::getType());
```
Comment (stonebuzz) [inline src/Inventory/Conf.php:971]:
```suggestion
            echo \State::createTabEntry(__('Status to apply'), 0, \State::getType());
```

File: src/Inventory/Conf.php [hunk 11]
```diff
@@ -935,38 +935,51 @@ public function showConfigForm()
...
             echo "</td>";
             echo "</tr>";
             //blocaction with status
-            echo "<tr class='tab_bg_1'><td colspan=2></td>";
+            echo "<tr class='tab_bg_1' style='display:none' id='bloc_status_action1'><td colspan=2></td>";
             echo "<td>";
-            echo "<span id='blocaction1' style='display:none'>";
-            echo \State::createTabEntry(__('Change the status'), 0, \State::getType());
-            echo "</span>";
+            echo \State::createTabEntry(__('Status to change (all if empty)'), 0, \State::getType());
```
Comment (stonebuzz) [inline src/Inventory/Conf.php:954]:
```suggestion
            echo \State::createTabEntry(__('If the asset status is (all if empty)'), 0, \State::getType());
```

File: src/Inventory/Conf.php (hunk 10 above)
Comment (stonebuzz) [reply src/Inventory/Conf.php:971]: Can you display this configuration before the other one?

File: tests/functional/Agent.php [hunk 12]
```diff
@@ -500,5 +500,112 @@ public function testStaleActions()
...
+        $this->integer($states_id2)->isGreaterThan(0);
+
+        //set last agent contact far ago
+        $DB->update(
+            \Agent::getTable(),
+            ['last_contact' => date('Y-m-d H:i:s', strtotime('-1 year'))],
+            ['id' => $agent['id']]
+        );
+
+        //define sale agents actions
```
Comment (stonebuzz) [inline tests/functional/Agent.php:515]:
```suggestion
        //define stale agents actions
```

File: tests/functional/Agent.php [hunk 13]
```diff
@@ -500,5 +500,112 @@ public function testStaleActions()
...
+        );
+
+        //run crontask
+        $task = new \CronTask();
+        $this->integer(\Agent::cronCleanoldagents($task))->isIdenticalTo(1);
+
+        //check item has been updated
+        $this->boolean($item->getFromDB($item->fields['id']))->isTrue();
+        $this->integer($item->fields['is_deleted'])->isIdenticalTo(1);
+        $this->integer($item->fields['states_id'])->isIdenticalTo($states_id3);
```
Comment (stonebuzz) [inline tests/functional/Agent.php:609]: perhaps a final test to check that an asset with a status of "xxx" is not changed when the out-of-date agent configuration needs to

change the status of an asset with a status of "yyyy".

**Review by stonebuzz — CHANGES_REQUESTED** [review]

**Review by trasher — APPROVED** [review]
Not tested, but seems correct

File: src/Inventory/Conf.php [hunk 14]
```diff
@@ -1179,6 +1192,7 @@ public static function getDefaults(): array
             'stale_agents_delay'             => 0,
             'stale_agents_action'            => exportArrayToDB([0]),
             'stale_agents_status'            => 0,
+            'previous_stale_agents_status'        => exportArrayToDB([0]),
```
Comment (stonebuzz) [inline src/Inventory/Conf.php:1195]: can you fix indentation

File: src/Inventory/Conf.php [hunk 15]
```diff
@@ -935,38 +935,51 @@ public function showConfigForm()
...
             echo "<td>";
-            echo "<span id='blocaction1' style='display:none'>";
-            echo \State::createTabEntry(__('Change the status'), 0, \State::getType());
-            echo "</span>";
+            echo \State::createTabEntry(__('If the asset status is (all if empty)'), 0, \State::getType());
+            echo "</td>";
+            echo "<td width='20%'>";
+            Dropdown::showFromArray(
+                'previous_stale_agents_status',
+                State::getStateList(),
```
Comment (stonebuzz) [inline src/Inventory/Conf.php:959]: Why not put the list of statutes applicable to the inventory items directly ?
Rather than excluding other statuses with ```used```.

File: src/State.php (hunk 7 above)
Comment (stonebuzz) [inline src/State.php:658]: Useless (IMHO) see previous comment

**Review by stonebuzz — CHANGES_REQUESTED** [review]

File: src/Inventory/Conf.php (hunk 15 above)
Comment (stonebuzz) [reply src/Inventory/Conf.php]: fixed

File: src/State.php (hunk 7 above)
Comment (stonebuzz) [reply src/State.php]: fixed

File: src/Inventory/Conf.php (hunk 14 above)
Comment (stonebuzz) [reply src/Inventory/Conf.php]: fixed

File: src/Inventory/Conf.php [hunk 16]
```diff
@@ -1179,6 +1195,7 @@ public static function getDefaults(): array
             'stale_agents_delay'             => 0,
             'stale_agents_action'            => exportArrayToDB([0]),
             'stale_agents_status'            => 0,
+            'previous_stale_agents_status'   => exportArrayToDB([0]),
```
Comment (cedric-anne) [inline src/Inventory/Conf.php:1198]: Maybe `stale_agents_status_condition` would be more understandable.

At first, I thought you was trying to introduce some kind of history on the `stale_agents_status` values.

File: src/Inventory/Conf.php [hunk 17]
```diff
@@ -935,38 +935,54 @@ public function showConfigForm()
...
-            echo "</span>";
+            echo \State::createTabEntry(__('If the asset status is (all if empty)'), 0, \State::getType());
             echo "</td>";
             echo "<td width='20%'>";
-            echo "<span id='blocaction2' style='display:none'>";
+            $condition = [];
+            foreach ($CFG_GLPI['inventory_types'] as $inv_type) {
+                $condition['is_visible_' . strtolower($inv_type)] = 1;
+            }
             State::dropdown(
```
Comment (cedric-anne) [inline src/Inventory/Conf.php:961]: IMHO, this dropdown should display an `All` option (using the `toadd` param) and when it is selected, the value saved in db should be `["all"]`. This should also be the default value, instead of `[0]`.

Without this `All` option, a user will not able to restore the default behaviour, and, also, a user that opens the form to change another value and save it without changing the current dropdown selection will have the `[]` value stored and it will probably disable the feature.

**Review by cedric-anne — CHANGES_REQUESTED** [review]

File: src/Inventory/Conf.php [hunk 18]
```diff
@@ -935,38 +935,54 @@ public function showConfigForm()
...
-                $('#blocaction1').hide();
-                $('#blocaction2').hide();
+                function changestatus() {
+                    if ($('#dropdown_stale_agents_action$rand').val() != 0 && $('#dropdown_stale_agents_action$rand').val() != 2) {
+                        $('#bloc_status_action1').show();
+                        $('#bloc_status_action2').show();
+                    } else {
+                        $('#bloc_status_action1').hide();
+                        $('#bloc_status_action2').hide();
+                    }
```
Comment (AdrienClairembault) [inline src/Inventory/Conf.php:945]: Code can be simplified a bit

```suggestion
                    const display = $('#dropdown_stale_agents_action$rand').val() != 0 
                        && $('#dropdown_stale_agents_action$rand').val() != 2;
                       
                    $('#bloc_status_action1').toggle(display);
                    $('#bloc_status_action2').toggle(display);
```

File: src/Agent.php [hunk 19]
```diff
@@ -881,17 +881,31 @@ public static function cronCleanoldagents($task = null)
...
-                            //change status of agents linked assets
-                            $input = [
-                                'id'        => $item->fields['id'],
-                                'states_id' => $config['stale_agents_status'],
-                                'is_dynamic' => 1
-                            ];
-                            if ($item->update($input)) {
-                                $task->addVolume(1);
-                                $total++;
+                            $shouldUpdate = false;
```
Comment (AdrienClairembault) [inline src/Agent.php:884]: By convention GLPI use snake case for its variables.

It is not always enforced and you don't really need to fix it here but keep it in mind for futures PR as we try to keep things consistant everywhere if possible.

File: tests/functional/Agent.php [hunk 20]
```diff
@@ -500,5 +500,124 @@ public function testStaleActions()
         $this->boolean($item->getFromDB($item->fields['id']))->isTrue();
         $this->integer($item->fields['is_deleted'])->isIdenticalTo(1);
         $this->integer($item->fields['states_id'])->isIdenticalTo($states_id);
+
+        //create another new status
+        $states_id2 = $state->add(['name' => 'Stale2']);
+        $this->integer($states_id2)->isGreaterThan(0);
```
Comment (AdrienClairembault) [inline tests/functional/Agent.php:506]: You can use the `createItem` method to simplify your tests:

```suggestion
        $state2 = $this->createItem(\State::class, ['name' => 'Stale2']);
        $states_id2 = $state2->getID();
```

It do the `->isGreaterThan(0)` part for you and will also run additional tests to ensure all supplied values were set correctly.

File: tests/functional/Agent.php [hunk 21]
```diff
@@ -500,5 +500,124 @@ public function testStaleActions()
...
+        $this->integer(\Agent::cronCleanoldagents($task))->isIdenticalTo(1);
+
+        //check item has been updated
+        $this->boolean($item->getFromDB($item->fields['id']))->isTrue();
+        $this->integer($item->fields['is_deleted'])->isIdenticalTo(1);
+        $this->integer($item->fields['states_id'])->isIdenticalTo($states_id2);
+
+        //create another new status for test while previous status is empty (0)
+        $states_id3 = $state->add(['name' => 'Stale3']);
+        $this->integer($states_id3)->isGreaterThan(0);
```
Comment (AdrienClairembault) [inline tests/functional/Agent.php:543]: Same here

File: tests/functional/Agent.php [hunk 22]
```diff
@@ -500,5 +500,124 @@ public function testStaleActions()
...
+        $this->integer(\Agent::cronCleanoldagents($task))->isIdenticalTo(1);
+
+        //check item has been updated
+        $this->boolean($item->getFromDB($item->fields['id']))->isTrue();
+        $this->integer($item->fields['is_deleted'])->isIdenticalTo(1);
+        $this->integer($item->fields['states_id'])->isIdenticalTo($states_id3);
+
+        //create another new status for test while previous status is empty (0)
+        $states_id4 = $state->add(['name' => 'Stale4']);
+        $this->integer($states_id4)->isGreaterThan(0);
```
Comment (AdrienClairembault) [inline tests/functional/Agent.php:570]: Same here

**Review by AdrienClairembault — CHANGES_REQUESTED** [review]
Code looks good beside some minors requested changes.

I'm not familiar enough with this feature to test its functional side.

File: src/Inventory/Conf.php (hunk 16 above)
Comment (stonebuzz) [reply src/Inventory/Conf.php:1198]: fixed

File: src/Inventory/Conf.php (hunk 17 above)
Comment (stonebuzz) [reply src/Inventory/Conf.php:961]: fixed

File: src/Inventory/Conf.php (hunk 18 above)
Comment (stonebuzz) [reply src/Inventory/Conf.php:945]: fixed

File: src/Agent.php (hunk 19 above)
Comment (stonebuzz) [reply src/Agent.php:884]: fixed

File: tests/functional/Agent.php (hunk 20 above)
Comment (stonebuzz) [reply tests/functional/Agent.php:506]: fixed

File: src/Inventory/Conf.php (hunk 17 above)
Comment (cedric-anne) [reply src/Inventory/Conf.php:963]: Somewhere in the process, the value is probably an escaped JSON string and could not be decoded.

```php
json_decode('["all"]'); // -> ["all"] (an array)
json_decode('[\"all\"]'); // -> "[\"all\"]" (a string)
```

Comment (stonebuzz) [reply src/Inventory/Conf.php:963]: Thanks for the lead I've found

see: [72a869e](https://github.com/glpi-project/glpi/pull/16512/commits/72a869ecba45443e7a702ad7f9d0c0a69c4427ae)

File: src/Inventory/Conf.php (hunk 18 above)
Comment (AdrienClairembault) [reply src/Inventory/Conf.php:945]: Are you sure ? Code seems unchanged

**Review by AdrienClairembault — APPROVED** [review]

File: src/Inventory/Conf.php (hunk 18 above)
Comment (stonebuzz) [reply src/Inventory/Conf.php]: yes (sorry) I modified the code proposal, as the ```select``` allows ```mutliple``` selection the ```val()``` returns an ```array```

**Review by trasher — APPROVED** [review]

**Review by cedric-anne — APPROVED** [review]
